storage/src/journal/contiguous/variable.rs

```diff
@@ -0,0 +1,1436 @@
+                    count += 1;
+                }
+                count
+            };
+
+            let size = (last_section * items_per_section) + items_in_last_section;
+            (oldest_pos, size)
+        };
+
+        let locations_size = locations.size().await?;
+        if locations_size > data_size {
+            // Locations ahead of data → should never happen (violates write ordering)
+            return Err(Error::Corruption(format!(
+                "locations ahead of data: locations_size={locations_size} > data_size={data_size}"
+            )));
+        }
+        // Apply any operations that are missing from the locations journal.
+        if locations_size < data_size {
+            Self::add_missing_locations(data, locations, locations_size).await?;
+        }
+
+        // Prune the locations journal to match the data journal if necessary.
+        // We prune data before locations so we should never need to catch data up to locations.
+        match locations.oldest_retained_pos().await? {
+            Some(oldest_retained_pos) if oldest_retained_pos > data_oldest_pos => {
+                // Locations pruned ahead of data → should never happen (violates write ordering)
+                return Err(Error::Corruption(format!(
+                    "locations pruned ahead of data: locations_oldest={oldest_retained_pos} > data_oldest={data_oldest_pos}"
+                )));
+            }
+            Some(oldest_retained_pos) if oldest_retained_pos < data_oldest_pos => {
+                // Locations behind on pruning → prune to catch up
+                locations.prune(data_oldest_pos).await?;
+            }
+            None if data_oldest_pos < data_size => {
+                // The locations journal returned `None` from oldest_retained_pos(), which means
+                // it's fully pruned (all items have been pruned, so oldest_retained_pos == size).
+                //
+                // However, the data journal still has un-pruned items.
+                //
+                // Example:
+                //   - Size: 100 (both journals agree we have 100 total items)
+                //   - Data oldest: 20 (data has items 20-99)
+                //   - Locations oldest: 100 (locations has pruned everything)
+                //
+                // This violates our invariant: data is always pruned BEFORE locations.
+                // If the data journal has items 20-99, the locations journal should too.
+                return Err(Error::Corruption(format!(
+                    "locations pruned ahead of data: locations fully pruned (oldest_retained_pos={data_size}) but data_oldest={data_oldest_pos}"
+                )));
+            }
+            _ => {
+                // Pruning is consistent
+            }
+        }
+
+        locations.sync().await?;
+        assert_eq!(locations.size().await?, data_size);
+        assert_eq!(
+            locations.oldest_retained_pos().await?,
```

Comment:
Will this return `None` if it is an empty journal?

I think that may cause an unnecessary assertion failure?

Reply:
Ah, interesting. Missed that.